src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs

```diff
@@ -204,24 +197,54 @@ private DependencyGraphSpec ReadProjectDependencyGraph(PackageReferenceArgs pack
             // If the user did not specify a version then write the exact resolved version
             if (packageReferenceArgs.NoVersion)
             {
-                // Get the flattened restore graph
-                var flattenedRestoreGraph = restorePreviewResult
-                    .Result
-                    .RestoreGraphs
-                    .First()
-                    .Flattened;
-
-                // Get the package entry and version from the graph
-                var packageEntry = flattenedRestoreGraph
-                    .Single(p => p.Key.Name.Equals(packageReferenceArgs.PackageDependency.Id, StringComparison.OrdinalIgnoreCase));
-                var resolvedVersion = packageEntry
-                    .Key
-                    .Version;
-
-                //Update the packagedependency with the new version
-                packageReferenceArgs.PackageDependency = new PackageDependency(packageReferenceArgs.PackageDependency.Id,
-                    VersionRange.Parse(resolvedVersion.ToString()));
+                // Get the package version from the graph
+                var resolvedVersion = privateGetPackageVersionFromRestoreResult(restorePreviewResult, packageReferenceArgs);
+
+                if (resolvedVersion != null)
+                {
+                    //Update the packagedependency with the new version
+                    packageReferenceArgs.PackageDependency = new PackageDependency(packageReferenceArgs.PackageDependency.Id,
+                        VersionRange.Parse(resolvedVersion.ToString()));
+                }
+            }
+        }
+
+        private NuGetVersion privateGetPackageVersionFromRestoreResult(RestoreResultPair restorePreviewResult,
+            PackageReferenceArgs packageReferenceArgs)
+        {
+            // Get the restore graphs from the restore result
+            var restoreGraphs = restorePreviewResult
+                .Result
+                .RestoreGraphs;
+
+            if (packageReferenceArgs.Frameworks?.Any() == true)
+            {
+                // If the user specified frameworks then we get the flattened graphs  only from the compatible frameworks.
+                var userSpecifiedFrameworks = new HashSet<NuGetFramework>(
+                    packageReferenceArgs
+                    .Frameworks
+                    .Select(f => NuGetFramework.Parse(f)));
+
+                restoreGraphs = restoreGraphs
+                    .Where(r => userSpecifiedFrameworks.Contains(r.Framework));
```

Comment:
What if the user accidentally specified a framework that doesn't exist in the project? a error message for that scenario might help, looks like it would just go on and possibly fail with an unrelated error message